futures-select-macro/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+                },
+            }
+        }
+
+        Ok(select)
+    }
+}
+
+// Enum over all the cases in which the `select!` waiting has completed and the result
+// can be processed.
+//
+// `enum __PrivResult<_1, _2, ...> { _1(_1), _2(_2), ..., Complete }`
+fn declare_result_enum(
+    result_ident: Ident,
+    variants: usize,
+    complete: bool,
+    span: Span
+) -> (Vec<Ident>, syn::ItemEnum) {
+    // "_0", "_1", "_2"
+    let variant_names: Vec<Ident> =
+        (0..variants)
+            .map(|num| Ident::new(&format!("_{}", num), span))
+            .collect();
+
+    // `<_1, _2, ...>`
+    let enum_generics = variant_names.iter()
+        .map(|name| {
+            syn::GenericParam::Type(parse_quote!( #name ))
+        })
+        .collect();
+
+    let unit_variant = |name| syn::Variant {
+        attrs: vec![],
+        ident: Ident::new(name, span),
+        fields: syn::Fields::Unit,
+        discriminant: None,
+    };
+
+    // `_1(_1), _2(_2), ..., Complete`
+    let enum_variants = variant_names.iter()
+        .map(|name| parse_quote!( #name ( #name ) ))
+        .chain(if complete { Some(unit_variant("Complete")) } else { None })
+        .collect();
+
+    (
+        variant_names,
+        syn::ItemEnum {
+            attrs: vec![],
+            vis: syn::Visibility::Inherited,
+            enum_token: syn::token::Enum::default(),
+            ident: result_ident,
+            generics: syn::Generics {
+                lt_token: None,
+                params: enum_generics,
+                gt_token: None,
+                where_clause: None,
+            },
+            brace_token: syn::token::Brace { span },
+            variants: enum_variants,
+        },
+    )
+}
+
+/// The `select!` macro.
+#[proc_macro_hack]
+pub fn select(input: TokenStream) -> TokenStream {
+    let parsed = syn::parse_macro_input!(input as Select);
+
+    let futures_crate: syn::Path = parsed.futures_crate_path.unwrap_or_else(|| parse_quote!(::futures_util));
+    let rand_crate: syn::Path = parse_quote!(#futures_crate::rand_reexport);
+
+    // should be def_site, but that's unstable
+    let span = Span::call_site();
+
+    let enum_ident = Ident::new("__PrivResult", span);
+
+    let (variant_names, enum_item) = declare_result_enum(
+        enum_ident.clone(),
+        parsed.normal_fut_exprs.len(),
+        parsed.complete.is_some(),
+        span,
+    );
+
+    // bind non-`Ident` future exprs w/ `let`
+    let mut future_let_bindings = Vec::with_capacity(parsed.normal_fut_exprs.len());
+    let bound_future_names: Vec<_> = parsed.normal_fut_exprs.into_iter()
+        .zip(variant_names.iter())
+        .map(|(expr, variant_name)| {
+            match expr {
+                // Don't bind futures that are already a path.
+                // This prevents creating redundant stack space
+                // for them.
+                syn::Expr::Path(path) => path,
+                x => {
```

Comment:
I would either call this `other` instead of `x`, or use `_` here and interpolate as `#expr` below.